### ISSUE TITLE
label_flag transition crash after shutdown when using string

### ISSUE THREAD

**Issue by moroten** [bazelbuild/bazel#10500]
### Description of the problem:

Bazel crashes when not using `Label()` for values in a transition of `label_flag`s. Should both be allowed or just one way?

Note that the crash depends on the Bazel server state. How is that even possible?

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
touch WORKSPACE

cat <<EOF > BUILD.bazel
load(":rules.bzl", "rule_with_transition")
package(default_visibility = ["//visibility:public"])
label_flag(
    name = "the_label_flag",
    build_setting_default = ":bad_filegroup",
)
filegroup(name = "bad_filegroup", srcs = [])
filegroup(name = "good_filegroup", srcs = [])

rule_with_transition(
    name = "output_target",
    srcs = [":the_label_flag"],
)
EOF

cat <<EOF > rules.bzl
def _my_transition_impl(settings, attr):
    print(settings)
    return {
        # This line works.
        "//:the_label_flag": Label("//:good_filegroup"),
        # To trigger a crash, run 'bazel shutdown' and replace with
        # "//:the_label_flag": "//:good_filegroup",
    }

my_transition = transition(
    implementation = _my_transition_impl,
    inputs = ["//:the_label_flag"],
    outputs = ["//:the_label_flag"],
)

def _rule_impl(ctx):
    print(ctx.attr.srcs)

rule_with_transition = rule(
    implementation = _rule_impl,
    # cfg = my_transition,
    attrs = {
        "srcs": attr.label_list(allow_files = True, cfg = my_transition),
        "_whitelist_function_transition": attr.label(default = "@bazel_tools//tools/whitelists/function_transition_whitelist"),
    },
)
EOF
```

Now run
```
bazel build output_target  # OK
# Replace Label("xyz") with "xyz" in my_transition().
sed -i 's/Label(\(.*\))/\1/g' rules.bzl
bazel build output_target  # OK
bazel shutdown
bazel build output_target  # CRASH (logs below)
```

### What operating system are you running Bazel on?

Linux

### What's the output of `bazel info release`?

`release 2.0.0`

###  Have you found anything relevant by searching the web?

Nothing found when searching for `label_flag` on GitHub issues: https://github.com/bazelbuild/bazel/issues?utf8=%E2%9C%93&q=is%3Aissue+label_flag

Relates to bazel-dicsuss [platforms, build_settings, transitions and features](https://groups.google.com/forum/#!topic/bazel-discuss/EJIk-SViZ7Y).

The example above is loosely based on https://github.com/bazelbuild/bazel/blob/master/src/test/shell/integration/starlark_configurations_test.sh.

### Any other information, logs, or outputs that you want to share?

Trace back from the crash:
```
Starting local Bazel server and connecting to it...
INFO: Elapsed time: 3.214s
INFO: 0 processes.
java.lang.RuntimeException: Unrecoverable error while evaluating node '//:the_label_flag BuildConfigurationValue.Key[d41cc611c4c1018165e95e4627a05585fa8ab8de6e32505ca046e28f38a3dc95] false' (requested by nodes '//:output_target BuildConfigurationValue.Key[419a3f9c03a725a237290b10359c9923007948fc3772740ed08cd228e37927c7] false')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:531)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class com.google.devtools.build.lib.cmdline.Label (java.lang.String is in module java.base of loader 'bootstrap'; com.google.devtools.build.lib.cmdline.Label is in unnamed module of loader 'app')
	at com.google.devtools.build.lib.rules.LabelBuildSettings.lambda$static$0(LabelBuildSettings.java:63)
	at com.google.devtools.build.lib.packages.Attribute$SimpleLateBoundDefault.resolve(Attribute.java:1594)
	at com.google.devtools.build.lib.packages.Attribute$LabelLateBoundDefault.resolve(Attribute.java:1725)
	at com.google.devtools.build.lib.analysis.DependencyResolver.resolveLateBoundDefault(DependencyResolver.java:574)
	at com.google.devtools.build.lib.analysis.DependencyResolver.resolveAttributes(DependencyResolver.java:529)
	at com.google.devtools.build.lib.analysis.DependencyResolver.visitRule(DependencyResolver.java:438)
	at com.google.devtools.build.lib.analysis.DependencyResolver.dependentNodeMap(DependencyResolver.java:276)
	at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.computeDependencies(ConfiguredTargetFunction.java:561)
	at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:310)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:454)
FAILED: Build did NOT complete successfully (5 packages loaded, 6 targets configured)
```

**Comment by gregestren** [comment]
I can easily reproduce this.

**Comment by gregestren** [comment]
I have an incoming fix.

I had to dig around a bit to figure out what even makes sense (i.e. support the above example or fail with a cleaner error?).

I chose to support it based on the logic that all other [flag types](https://docs.bazel.build/versions/master/skylark/lib/config.html) are considered valid if they pass [Type.convert](https://github.com/bazelbuild/bazel/blob/d21a0d194b3e83ca585e466ed4f6a383776bfb76/src/main/java/com/google/devtools/build/lib/packages/Type.java#L78).

Label types are unique in that they accept both Label and String types, while int requires actual ints (no `"15"`), booleans actual booleans (no `"true"`), etc. But that's a property of the Label type: the generic logic is still the same.
